### PR TITLE
feat: support for detaching subscriptions

### LINKED ISSUE #1031: Add support for detaching subscriptions
The server API is not released yet, but we need to add client support for detaching subscriptions, as well as the "detached" flag for subscription metadata.

https://github.com/googleapis/googleapis/blob/e45a2f54dce1ac68497acf305145f9e5bb7e1e48/google/pubsub/v1/pubsub.proto#L130

### PR DESCRIPTION
**Pull request by feywind**
This is not yet available for public service use, but client library support is in progress.

PR is a work in progress.

Fixes: https://github.com/googleapis/nodejs-pubsub/issues/1031

### CODE REVIEW

File: samples/detachSubscription.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      `Subscription ${subscriptionName} detached status: ${metadata.detached}`
+    );
+
+    await sub.detach();
+    console.log(`Subscription ${subscriptionName} detached.`);
+
+    const updatedMetadata = await sub.getMetadata()[1];
+    console.log(
+      `Subscription ${subscriptionName} detached status: ${updatedMetadata.detached}`
+    );
```
Comment (anguillanneuf) [inline samples/detachSubscription.js:57]: How about 1. call `.detach()` 2. if detached, print out successfully detached, else not detached?

File: samples/detachSubscription.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+//   title: Detach Subscription
+//   description: Detaches a subscription from a topic.
+//   usage: node detachSubscription.js <subscription-name>
+
+function main(subscriptionName = 'YOUR_SUBSCRIPTION_NAME') {
+  // [START pubsub_detach_subscription]
+  /**
+   * TODO(developer): Uncomment these variables before running the sample.
+   */
+  // const subscriptionName = 'YOUR_SUBSCRIPTION_NAME';
```
Comment (anguillanneuf) [inline samples/detachSubscription.js:35]: Add a comment - provide an existing subscription.

File: samples/detachSubscription.js (hunk 1 above)
Comment (feywind) [reply samples/detachSubscription.js:57]: Yeah, what I was trying to demonstrate here is the ability to check the "detached" flag from metadata. But it seems like both are useful. And I do want to add a helper that does the metadata check, because it looks gross as it is.

**Comment by feywind** [comment]
Seems like codecov is annoyed because the system-test doesn't cover the new sample. (Which is true, but I was just planning to make a second PR that we can merge when it's ready.)

**Comment by feywind** [comment]
@anguillanneuf @pradn Do y'all mind giving this a look-over?

File: samples/detachSubscription.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+// limitations under the License.
+
+/**
+ * This sample demonstrates how to detach subscriptions with the
+ * Google Cloud Pub/Sub API.
+ *
+ * For more information, see the README.md under /pubsub and the documentation
+ * at https://cloud.google.com/pubsub/docs.
+ *
+ * Note: This sample is NOT currently included in the system-test, because
```
Comment (JustinBeckwith) [inline samples/detachSubscription.js:22]: Should we put a `DO NOT MERGE` on this until that's fixed?

File: samples/detachSubscription.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    console.log(
+      `Subscription ${subscriptionName} 'after' detached status: ${updatedDetached}`
+    );
+  }
+
+  detachSubscription();
+  // [END pubsub_detach_subscription]
+}
+
+process.on('unhandledRejection', err => {
```
Comment (JustinBeckwith) [inline samples/detachSubscription.js:67]: I believe this is the stuff we're looking for now:
```js

main(...process.argv.slice(2)).catch(err => {
  console.error(err.message);
  process.exitCode = 1;
});
process.on('unhandledRejection', (err) => {
  console.error(err.message);
  process.exitCode = 1;
});
```

File: src/pubsub.ts [hunk 5]
```diff
@@ -541,32 +546,125 @@ export class PubSub {
...
   }
+
+  detachSubscription(
+    name: string,
+    gaxOpts?: CallOptions
+  ): Promise<DetachSubscriptionResponse>;
+  detachSubscription(name: string, callback: DetachSubscriptionCallback): void;
+  detachSubscription(
+    name: string,
+    gaxOpts: CallOptions,
```
Comment (JustinBeckwith) [inline src/pubsub.ts:557]: Usually if `gaxOpts` is optional in one invocation model (promises), it can be optional in the callback model as well.  little weird that you MUST specify `gaxOpts` only for callbacks.

Side crazy question - what if we just like didn't support callbacks for new APIs? May be a good way to try this out and measure the response.

File: src/pubsub.ts [hunk 6]
```diff
@@ -541,32 +546,125 @@ export class PubSub {
...
+   * @param {object} [gaxOpts] Request configuration options, outlined
+   *     here: https://googleapis.github.io/gax-nodejs/interfaces/CallOptions.html.
+   * @param {DetachSubscriptionCallback} [callback] Callback function.
+   * @returns {Promise<DetachSubscriptionResponse>}
+   *
+   * @example
+   * const {PubSub} = require('@google-cloud/pubsub');
+   * const pubsub = new PubSub();
+   *
+   * pubsub.detachSubscription('my-sub', function(err, topic, apiResponse) {
```
Comment (JustinBeckwith) [inline src/pubsub.ts:575]: Maybe use an arrow function?

File: src/pubsub.ts [hunk 7]
```diff
@@ -541,32 +546,125 @@ export class PubSub {
...
+        this.options.port = undefined;
+      }
+    } else {
+      this.options.port = parseInt(baseUrlParts[1], 10);
+    }
+
+    // If this looks like a GCP URL of some kind, don't go into emulator
+    // mode. Otherwise, supply a fake SSL provider so a real cert isn't
+    // required for running the emulator.
+    const officialUrlMatch = this.options.servicePath!.match(
```
Comment (JustinBeckwith) [inline src/pubsub.ts:660]: `this.options.servicePath!.endsWith('.googleapis.com')` is so readable for regex-impaired folk like me

File: src/pubsub.ts [hunk 8]
```diff
@@ -541,32 +546,125 @@ export class PubSub {
...
+      callback!
+    );
+  }
+
   /**
    * Determine the appropriate endpoint to use for API requests, first trying
-   * the local `apiEndpoint` parameter. If the `apiEndpoint` parameter is null
-   * we try Pub/Sub emulator environment variable (PUBSUB_EMULATOR_HOST),
-   * otherwise the default JSON API.
+   * the local `apiEndpoint` parameter. If the `apiEndpoint` parameter is null,
```
Comment (JustinBeckwith) [inline src/pubsub.ts:614]: This whole section below feels disconnected from the rest of the content of the PR.  Should these be two separate PRs?

Comment (feywind) [reply src/pubsub.ts:614]: Fair. I needed this to test the rest of the PR, but this part could be merged first.

File: src/pubsub.ts (hunk 6 above)
Comment (feywind) [reply src/pubsub.ts:575]: I was mostly trying to match existing examples in the comments, but I do agree it should be an arrow function.

File: src/pubsub.ts (hunk 5 above)
Comment (feywind) [reply src/pubsub.ts:557]: Yeah, I was again just matching existing style. There was some discussion about dropping callbacks earlier, and I think that was unpopular. I'm up for discussing it though.

File: samples/detachSubscription.js (hunk 4 above)
Comment (feywind) [reply samples/detachSubscription.js:67]: So, I'm slightly concerned there is Zalgo in that ordering. (This objection is a bit ironic since I helped author that sample style update. >_>) That can be a separate discussion though.

File: samples/detachSubscription.js (hunk 3 above)
Comment (feywind) [reply samples/detachSubscription.js:22]: The other languages seem to be doing that, so I might shift that way, yeah.

File: samples/detachSubscription.js (hunk 4 above)
Comment (JustinBeckwith) [reply samples/detachSubscription.js:67]: Ugh.  @bcoe @sofisl @fhinkel

File: samples/detachSubscription.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+  async function detachSubscription() {
+    // Gets the status of the existing subscription
+    const sub = pubSubClient.subscription(subscriptionName);
+    const [detached] = await sub.detached();
+    console.log(
+      `Subscription ${subscriptionName} 'before' detached status: ${detached}`
+    );
+
+    await pubSubClient.detachSubscription(subscriptionName);
```
Comment (anguillanneuf) [inline samples/detachSubscription.js:54]: Does `detachSubscription` throw explicit errors?  If so, we could list them in the sample.

Comment (feywind) [reply samples/detachSubscription.js:54]: Do you mean things like "the subscription was already detached"? I'm actually not sure.

All of those gRPC calls are theoretically capable of throwing a bunch of gRPC errors, and the other samples don't call those out, so I'd want to keep everything as harmonized across samples there as I can.

File: samples/detachSubscription.js (hunk 3 above)
Comment (feywind) [reply samples/detachSubscription.js:22]: Updating this from internal discussions: the feature will be available to allow-listed test projects, so we can go ahead with merging the full set in. I'll re-enable this.

File: src/pubsub.ts (hunk 5 above)
Comment (feywind) [reply src/pubsub.ts:557]: Hmm, I see what was going on here:
```
  detachSubscription(
    name: string,
    gaxOpts?: CallOptions
  ): Promise<DetachSubscriptionResponse>;
  detachSubscription(name: string, callback: DetachSubscriptionCallback): void;
  detachSubscription(
    name: string,
    gaxOpts: CallOptions,
    callback: DetachSubscriptionCallback
  ): void;
  detachSubscription(
    name: string,
    optsOrCallback?: CallOptions | DetachSubscriptionCallback,
    callback?: DetachSubscriptionCallback
  ): Promise<DetachSubscriptionResponse> | void {
```

Basically this is just TypeScript resolution shenanigans. If you call with only gaxOpts, you want it to give you a Promise. If you call with gaxOpts and a callback, you get the second variant. If you call with only a callback, it will fall through to the last one, which has optsOrCallback. I don't think it has an effect one way or another on the runtime behavior, but it's sort of gross... but I get why Dave did it that way, too.

My "want a pony" goals for the upcoming quarters include switching everything to use Promise internally, and wrap stuff with callbackify(). There are a few things I need to figure out first though. (Like methods that just pass through callbacks directly.)

File: samples/detachSubscription.js (hunk 9 above)
Comment (anguillanneuf) [reply samples/detachSubscription.js:54]: I was referring to explicit exceptions. Java/Python samples are generally against generic exceptions so I was wondering about Node.

```python
    try:
        publisher_client.detach_subscription(subscription_path)
    except (GoogleAPICallError, RetryError, ValueError, Exception) as err:
        print(err)
```

Comment (feywind) [reply samples/detachSubscription.js:51]: Ah, I see. Yeah, the samples are all written in JavaScript, rather than TypeScript, so there's basically no rigor anywhere in them, outside of catching exceptions. I don't think even TypeScript has any real concept of typed exceptions. You just basically have to catch all the exceptions and filter for what you want. We could put a comment on it showing what exceptions might happen, but it would be very hard to figure that out for sure, and I'm afraid of it getting out of sync (e.g. if gRPC changed their exceptions thrown, and the Pub/Sub samples don't get updated).

**Comment by feywind** [comment]
The system test runs locally for me, and the log file for Kokoro seems to be blank. But this reminds me that this PR probably also needs a system test, anyway.

File: samples/detachSubscription.js (hunk 9 above)
Comment (anguillanneuf) [reply samples/detachSubscription.js:51]: Thanks for explaining. I see now.

**Review by anguillanneuf — APPROVED** [review]
LGTM samples :)
